### PR TITLE
Chore: test simplification for stateful set

### PR DESCRIPTION
**Pull request by luoqr96**
simplify the tests that involved with stateful set in pkg/apis/apps/validation/validation_test.go

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Table-driven tests are awesome, but they can be very verbose and obscure what's actually being tested in each case.

This PR simplifies the usage of stateful set in pkg/apis/apps/validation/validation_test.go

#### Which issue(s) this PR fixes:

This PR only solves a small part of  #114511

#### Special notes for your reviewer:

Thank you for taking the time to review my first issue on GitHub! As this is my first time submitting an issue, I would greatly appreciate any feedback you may have on how I can improve my submission in the future.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/apis/apps/validation/validation_test.go [hunk 1]
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
 	return &intOrStr
 }
 
+type setTweak func(s apps.StatefulSet)
+
+func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweaks ...setTweak) apps.StatefulSet {
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:47]: `makeStatefulSet` or `mkStatefulSet` seem better - "Set" could be a replicase, daemonset, statefulset, ...

File: pkg/apis/apps/validation/validation_test.go [hunk 2]
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
 	return &intOrStr
 }
 
+type setTweak func(s apps.StatefulSet)
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:45]: maybe `tweakStatefulSet` ?

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:47]: I don't think you need a pointer to the map - just passing the map "by value" is sufficient.

But also, you could make a tweak func?  Set a sane default in the `makeStatefulSet` func, and call the tweak ONLY from the cases that need a specific selector (which emphasizes that the selector is important) like:

```
fun setSelector(labels map[string]string) tweakStatefulSet {
    return func(ss apps.StatefulSet) {
        ss.Spec.Selector = &metav1.LabelSelector{MatchLabels: &labels)}
    }
}
```

File: pkg/apis/apps/validation/validation_test.go [hunk 3]
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
...
+
+func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweaks ...setTweak) apps.StatefulSet {
+	s := apps.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+		Spec: apps.StatefulSetSpec{
+			PodManagementPolicy: apps.OrderedReadyPodManagement,
+			Selector:            &metav1.LabelSelector{MatchLabels: *selectorLabels},
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:52]: This should produce a fully validate-safe object - I feel like you're maybe missing fields?  Isn't replicas required?  I see it missing in some test cases, so maybe not.

File: pkg/apis/apps/validation/validation_test.go [hunk 4]
```diff
@@ -463,118 +467,49 @@ func TestValidateStatefulSet(t *testing.T) {
...
-					Selector:            &metav1.LabelSelector{MatchLabels: validLabels},
-					Template:            validPodTemplate.Template,
-					Replicas:            3,
-					UpdateStrategy: apps.StatefulSetUpdateStrategy{Type: apps.OnDeleteStatefulSetStrategyType,
-						RollingUpdate: func() *apps.RollingUpdateStatefulSetStrategy {
-							return &apps.RollingUpdateStatefulSetStrategy{Partition: 1}
-						}()},
-				},
-			},
+			set:  makeSet(&validLabels, &validPodTemplate, tweakRepicas(3), tweakUpdateStrategy(makeUpdateStrategy(tweakUpdateStrategyType(apps.OnDeleteStatefulSetStrategyType), tweakRollingUpdatePartition(1)))),
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:484]: You might want to linewrap these cases of multiple tweaks, like:

```
set:  makeSet(&validLabels, &validPodTemplate,
     tweakRepicas(3), 
tweakUpdateStrategy(
    makeUpdateStrategy(
        tweakUpdateStrategyType(apps.OnDeleteStatefulSetStrategyType),
        tweakRollingUpdatePartition(1)
))),
```

or something - the goal is to call out what makes this case unique

**Review by thockin — COMMENTED** [review]
What a great cleanup!

**Comment by luoqr96** [comment]
Thank you! I'll simplify all the usage of stateful set in validation_test.go ASAP> What a great cleanup!

File: pkg/apis/apps/validation/validation_test.go (hunk 2 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:45]: modified as suggested

File: pkg/apis/apps/validation/validation_test.go (hunk 1 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:47]: modified as suggested

Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:47]: modified as suggested

File: pkg/apis/apps/validation/validation_test.go (hunk 3 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:52]: Replicas occur in 17/74 stateful set test cases, so I think maybe it is better to create a tweak func for it other than set a default value.

File: pkg/apis/apps/validation/validation_test.go (hunk 4 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:484]: modified as suggested

**Comment by luoqr96** [comment]
@thockin Modified as you suggested, please review again!

I simplified all the test cases of the stateful set in this new commit, PTAL :)

File: pkg/apis/apps/validation/validation_test.go [hunk 5]
```diff
@@ -62,53 +62,104 @@ func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweak
...
+func tweakAnnotations(annotations map[string]string) statefulSetTweak {
+	return func(ss apps.StatefulSet) {
+		ss.ObjectMeta.Annotations = annotations
 	}
 }
 
-func tweakPodManagementPolicy(policy apps.PodManagementPolicyType) setTweak {
-	return func(s apps.StatefulSet) {
-		s.Spec.PodManagementPolicy = policy
+func tweakFinalizers(finalizers []string) statefulSetTweak {
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:89]: Optional:

For finalizers, labels, annotations, you accept maps or slices, which makes the caller have to specify them.  If the goal is to make the callers as obvious and readable as possible, there's another way to approach this.  YOu could do:

```
func tweakFinalizers(finalizers ...string) statefulSetTweak {
```

This let's the caller say `tweakFinalizers("finalizer1")` or, if they really need more than one, `tweakFinalizers("one", "two", "three")`.

For the maps, you could do either:

```
func tweakLabels(key, value) statefulSetTweak {
    return func(ss apps.StatefulSet) {
        if ss.Labels == nil {
            ss.Labels = map[string]string{}
        }
        ss.Labels[key] = value
    }
}
```

If we really need to set more than one, you can call this more than one time.  Or you can do something like:

```
func tweakMetaLabels(keysAndValues ...string) statefulSetTweak {
```

But this is sort of awkward, IMO

File: pkg/apis/apps/validation/validation_test.go [hunk 6]
```diff
@@ -62,53 +62,104 @@ func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweak
...
+	}
+}
+
+func tweakSelectorLabels(labels map[string]string) statefulSetTweak {
+	return func(ss apps.StatefulSet) {
+		ss.Spec.Selector = &metav1.LabelSelector{MatchLabels: labels}
+	}
+}
+
+func tweakRestartPolicy(rp api.RestartPolicy) statefulSetTweak {
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:119]: I'd include the word Template in the name here

File: pkg/apis/apps/validation/validation_test.go [hunk 7]
```diff
@@ -62,53 +62,104 @@ func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweak
...
 	}
 }
 
 type updateStrategyTweak func(s apps.StatefulSetUpdateStrategy)
 
-func makeUpdateStrategy(tweaks ...updateStrategyTweak) apps.StatefulSetUpdateStrategy {
+func mkUpdateStrategy(tweaks ...updateStrategyTweak) apps.StatefulSetUpdateStrategy {
 	s := apps.StatefulSetUpdateStrategy{
 		Type: apps.RollingUpdateStatefulSetStrategyType,
+		RollingUpdate: func() *apps.RollingUpdateStatefulSetStrategy {
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:160]: Why this `func` construct?  I'm missing the purpose of it.

File: pkg/apis/apps/validation/validation_test.go [hunk 8]
```diff
@@ -366,267 +364,169 @@ func TestValidateStatefulSet(t *testing.T) {
...
-					PodManagementPolicy: apps.ParallelPodManagement,
-					UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
-					Selector:            &metav1.LabelSelector{MatchLabels: validLabels},
-					Template:            validPodTemplate.Template,
-					Replicas:            3,
-					Ordinals:            &apps.StatefulSetOrdinals{Start: -2},
-				},
-			},
+			set: mkStatefulSet(&validPodTemplate,
+				tweakReplicas(3),
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:527]: This is an awesome example - is Replicas ACTUALLY important to this test case, or was it just set for no real reason?  It doesn't seem like it SHOULD matter.

This cleanup makes such things obvious, and now there can be followups to remove unnecessary noise  like this (if it is unnecessary :).  Followups, please - not the same PR.

**Review by thockin — COMMENTED** [review]
A few small comments.

This PR is now almost too big to review.  My brsain gets fatigued and will miss things at this size.  Ideally this would be a series of small commits:
1) add the basic "make" function and a couple of simple tweaks, use that to show the pattern
2...N) add one more tweak and use it in those cases that can now be converted (occasionally you have to add 2 tweaks to satisfy a case)
last) fix any obviously redundant or wrongly named cases 

Another asspect of this you might consider as a followup is adding the line-number to each test case (`line string // if not empty, will be logged with errors, use line() to set`) and a line() function:

```
// line returns the line number of the caller, if possible.  This is useful in
// tests with a large number of cases - when something goes wrong you can find
// which case more easily.
func line() string {                                                                                                                        
    _, _, line, ok := stdruntime.Caller(1)
    var s string
    if ok {
        s = fmt.Sprintf("%d", line) 
    } else {
        s = "<??>"
    }
    return s
}
```

And then print that in the tests  (e.g. in the "Run()" call).

See pkg/registry/core/service/storage/storage_test.go for example.  This can also be a followup or at least a different commit.

FINALLY (again as followup or as a final commit, all on its own) you can do whitespace cleanups, so things like:

```
	successCases := []testCase{
		{
			name:   "update replica count",
			old:    mkStatefulSet(&validPodTemplate),
			update: mkStatefulSet(&validPodTemplate, tweakReplicas(3)),
		},
		{
			name:   "update containers 1",
			old:    mkStatefulSet(&validPodTemplate),
			update: mkStatefulSet(addContainersValidTemplate),
		},
		{
			/// and so on
		},
	}
```

become:

```
	successCases := []testCase{{
		name:   "update replica count",
		old:    mkStatefulSet(&validPodTemplate),
		update: mkStatefulSet(&validPodTemplate, tweakReplicas(3)),
	}, {
		name:   "update containers 1",
		old:    mkStatefulSet(&validPodTemplate),
		update: mkStatefulSet(addContainersValidTemplate),
	}, {
		/// and so on
	}}
```

10K lines of tests will become 1K or less.  It's like magic.

File: pkg/apis/apps/validation/validation_test.go (hunk 6 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:119]: modified as suggested

File: pkg/apis/apps/validation/validation_test.go (hunk 7 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:160]: removed

**Comment by luoqr96** [comment]
Hi Hockin! I reorganized all the commits as suggested, PTAL!

I apologize for making a mess before.

@thockin

File: pkg/apis/apps/validation/validation_test.go (hunk 5 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go]: modified as suggested.

However, as the selector labels are typically already specified, I find it maybe more convenient to continue accepting a map in the tweak function.

Comment (thockin) [reply pkg/apis/apps/validation/validation_test.go]: OK - the goal is "as obvious as possible, and not one step further".  If you think that's the right tradeoff, go for it.

File: pkg/apis/apps/validation/validation_test.go [hunk 9]
```diff
@@ -944,24 +928,12 @@ func TestValidateStatefulSetUpdate(t *testing.T) {
...
-					PodManagementPolicy: apps.OrderedReadyPodManagement,
-					Selector:            &metav1.LabelSelector{MatchLabels: validLabels2},
-					Template:            validPodTemplate2.Template,
-					UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
-				},
-			},
+			old: mkStatefulSet(&validPodTemplate,
+				tweakSelectorLabels(validLabels),
+			),
+			update: mkStatefulSet(&validPodTemplate2,
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go:934]: single tweaks probably don't need linewraps.

File: pkg/apis/apps/validation/validation_test.go [hunk 10]
```diff
@@ -42,6 +42,38 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
...
+		Spec: apps.StatefulSetSpec{
+			PodManagementPolicy: apps.OrderedReadyPodManagement,
+			Selector:            &metav1.LabelSelector{MatchLabels: map[string]string{"a": "b"}},
+			Template:            template.Template,
+			UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
+		},
+	}
+
+	for _, tw := range tweaks {
+		tw(s)
```
Comment (thockin) [inline pkg/apis/apps/validation/validation_test.go]: This needs to be a pointer - I see you did that in the last commit, but each of these commits, as beautiful as they are, don't work?  Can you back-populate that into these commits?

File: pkg/apis/apps/validation/validation_test.go [hunk 11]
```diff
@@ -507,48 +504,27 @@ func TestValidateStatefulSet(t *testing.T) {
...
-					PodManagementPolicy: "foo",
-					Selector:            &metav1.LabelSelector{MatchLabels: validLabels},
-					Template:            validPodTemplate.Template,
-					Replicas:            3,
-					UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
-				},
-			},
+			set: mkStatefulSet(&validPodTemplate,
+				tweakPodManagementPolicy("foo"),
+				tweakReplicas(3),
```
[the remark this answers is not in the document]
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go]: Yes, I keep the replicas here as a literal translation. Would it be better to remove these unnecessary tweaks in a new commit or a new pull request?

File: pkg/apis/apps/validation/validation_test.go [hunk 12]
```diff
@@ -315,21 +324,9 @@ func TestValidateStatefulSet(t *testing.T) {
...
-				},
-				Spec: apps.StatefulSetSpec{
-					PodManagementPolicy: apps.OrderedReadyPodManagement,
-					Selector:            &metav1.LabelSelector{MatchLabels: validLabels},
-					Template:            validPodTemplate.Template,
-					UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
-				},
-			},
+			set: mkStatefulSet(&validPodTemplate,
+				tweakMetaLabels("NoUppercaseOrSpecialCharsLike=Equals", "bar"),
```
[the remark this answers is not in the document]
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go]: modified as suggested

File: pkg/apis/apps/validation/validation_test.go (hunk 9 above)
Comment (luoqr96) [reply pkg/apis/apps/validation/validation_test.go:929]: modified as suggested

**Comment by luoqr96** [comment]
Hi Hockin!

 I wanted to let you know that I have addressed all of the code comments you made on this pull request. I would appreciate it if you could take another look at the code to confirm that the changes meet your expectations. Thank you for your time and feedback.

By the way, I am not sure whether it would be better to remove unnecessary tweaks in a new commit or a new pull request. For now, I have left them unmodified as a literal translation.

@thockin

File: pkg/apis/apps/validation/validation_test.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (thockin) [reply pkg/apis/apps/validation/validation_test.go]: new PR is fine.  Let's make this PR "done" now and start a new one

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Comment by luoqr96** [comment]
Hi Hockin!

The CI shows the kubectl_request_timeout test failed. I pasted the log below.

```
[31mFAIL!
[5651](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/116722/pull-kubernetes-integration/1638399925064568832#1:build-log.txt%3A5651)
(Bmessage:NAME        READY   STATUS    RESTARTS   AGE
valid-pod   0/1     Pending   0          1s
I0322 04:48:53.102939   38649 streamwatcher.go:114] Unable to decode an event from the watch stream: context deadline exceeded
has not:Timeout
42 /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../test/cmd/request-timeout.sh
!!! [0322 04:48:53] Call tree:
!!! [0322 04:48:53]  1: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:47 run_kubectl_request_timeout_tests(...)
!!! [0322 04:48:53]  2: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:112 eVal(...)
!!! [0322 04:48:53]  3: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:141 juLog(...)
!!! [0322 04:48:53]  4: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:624 record_command(...)
!!! [0322 04:48:53]  5: hack/make-rules/test-cmd.sh:194 runTests(...)
```

However, this test passed in my local environment. I also pasted the corresponding logs below.

```
Successful
message:NAME        READY   STATUS    RESTARTS   AGE
valid-pod   0/1     Pending   0          0s
I0322 14:10:25.249865   45141 streamwatcher.go:114] Unable to decode an event from the watch stream: context deadline exceeded (Client.Timeout or context cancellation while reading body)
has:Timeout
```

Considering this PR is just a cleanup of pkg/apis/apps/validation/validation_test.go, I think the failure here is unrelated with this PR.

Could you please help me rerun this failed test? Or is there any bugs that I might have missed?

Thank you so much!

@thockin

**Comment by luoqr96** [comment]
Hi Hockin! I noticed that the test freeze has just started. Should I leave this PR as is and wait for the v1.28 release cycle?

@thockin

**Comment by thockin** [comment]
Yeah, it can't merge until the 1.28 window opens, but that's OK.  It's ready to go and it seems unlikely to grow a conflict in this area.

**Pull request merged**